### PR TITLE
refactor: decouple `Table` from `QueryContext`

### PR DESCRIPTION
**Pull request by dantengsky**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

part of 
- issue #6180

-----
- **Decouple `QueryContext` from `Table` by introducing trait `TableContext`**

   These seem to be one of the biggest knots in way of splitting `query` into sub-crates.
   
  `Catalog`, `Database`, `Table` and `TableContext` are extracted into **one new crate `common-catalog`** (since there are cyclic dependencies between them).
  
  only the interfaces are extracted. The implementations are kept in crate `query`, to keep this PR as small as possible, but they are likely to be able to extract from `query` now.
  
  Trait `TableContext` is a little bit coarse-grained, it could be split it into more dedicated taits later if necessary.
  
  Unit(integration) tests are not transferred yet. They will be addressed in subsequent PRs。
  
  
   
- **other on-demand extracted crates**
  - `common-config`
  - `common-settings`
     **the "semantics" of Settings has changed a little bit, with the assumption that the tenant will not be changed during the lifetime of a given session:**
     
     before this PR, the `Settings` hold a `SessionCtx`, and gives the tenant on demand by using `SessionCtx`
     https://github.com/datafuselabs/databend/pull/6665/files#diff-d2f273cc255b0aee531d4eb86b74f4f5df598bbd861389287630ca2b27f65ea9R65-R67
     
     in this PR, the tenant is acquired during the construction of `Settings`, and never changed afterward.
     
     https://github.com/datafuselabs/databend/pull/6665/files#diff-d2f273cc255b0aee531d4eb86b74f4f5df598bbd861389287630ca2b27f65ea9R219-R227
     
     this arrangement makes `Settings` no longer depend on `SessionCtx`

  - `common-fuse-meta`
      fuse meta and cache
  - `common-pipeline` (basic types)
     only the types that do not depends on `QueryContext` are extracted
  
Fixes #issue

### CODE REVIEW

File: common/catalog/src/catalog.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+    /// Table function
+
+    // Get function by name.
+    fn get_table_function(
+        &self,
+        _func_name: &str,
+        _tbl_args: TableArgs,
+    ) -> Result<Arc<dyn TableFunction>> {
+        unimplemented!()
```
Comment (zhang2014) [inline common/catalog/src/catalog.rs:160]: ErrorCode::Unimplemented is better.

**Review by zhang2014 — APPROVED** [review]
Dependencies are becoming complex. I'm not sure whether complex dependencies can solve slow links? But the changes is looks good to me.

File: common/catalog/src/catalog.rs (hunk 1 above)
Comment (dantengsky) [reply common/catalog/src/catalog.rs:160]: OK

File: query/src/interpreters/fragments/query_fragment_actions.rs [hunk 2]
```diff
@@ -29,6 +29,8 @@ use crate::api::ExecutePartialQueryPacket;
 use crate::api::FragmentPlanPacket;
 use crate::api::InitNodesChannelPacket;
 use crate::api::QueryFragmentsPlanPacket;
+use crate::clusters::ClusterHelper;
+use crate::sessions::query_ctx::TableContext;
```
Comment (BohuTANG) [inline query/src/interpreters/fragments/query_fragment_actions.rs:33]: `use crate::sessions::TableContext;` looks better?

Comment (dantengsky) [reply query/src/interpreters/fragments/query_fragment_actions.rs:33]: ok

Comment (BohuTANG) [reply query/src/interpreters/fragments/query_fragment_actions.rs:33]: Else LGTM

**Comment by BohuTANG** [comment]
To block Mergify merged to the main branch by 2-approves, I changed this PR to draft. Feel free to convert it to ready anytime after the comments are fixed. @dantengsky

**Comment by dantengsky** [comment]
> I'm not sure whether complex dependencies can solve slow links? 

It does not, in my local dev box, for the main crate, `databend-query`, no dramatic improvement found. 

but for the crates that have been extracted. it takes much lesser time to build. I think after this PR, more sub mods could be extracted, or easier to be extracted. and gradually, the main crate will benefit as well.

**Pull request merged**